tests/contrib/funsor/test_valid_models_enum.py

```diff
@@ -0,0 +1,444 @@
+        check_traceenum_requirements(tr_funsor, Trace())
+        try:
+            # coarser check: number of elements and squeezed shapes
+            for name, pyro_node in tr_pyro.nodes.items():
+                if pyro_node['type'] != 'sample':
+                    continue
+                funsor_node = tr_funsor.nodes[name]
+                assert pyro_node['packed']['log_prob'].numel() == funsor_node['log_prob'].numel()
+                assert pyro_node['packed']['log_prob'].shape == funsor_node['log_prob'].squeeze().shape
+                assert frozenset(f for f in pyro_node['cond_indep_stack'] if f.vectorized) == \
+                    frozenset(f for f in funsor_node['cond_indep_stack'] if f.vectorized)
+        except AssertionError:
+            for name, pyro_node in tr_pyro.nodes.items():
+                if pyro_node['type'] != 'sample':
+                    continue
+                funsor_node = tr_funsor.nodes[name]
+                pyro_packed_shape = pyro_node['packed']['log_prob'].shape
+                funsor_packed_shape = funsor_node['log_prob'].squeeze().shape
+                if pyro_packed_shape != funsor_packed_shape:
+                    err_str = "==> (dep mismatch) {}".format(name)
+                else:
+                    err_str = name
+                print(err_str, "Pyro: {} vs Funsor: {}".format(pyro_packed_shape, funsor_packed_shape))
+            raise
+
+    if _NAMED_TEST_STRENGTH >= 2:
+        try:
+            # medium check: unordered packed shapes match
+            for name, pyro_node in tr_pyro.nodes.items():
+                if pyro_node['type'] != 'sample':
+                    continue
+                funsor_node = tr_funsor.nodes[name]
+                pyro_names = frozenset(symbol_to_name[d] for d in pyro_node['packed']['log_prob']._pyro_dims)
+                funsor_names = frozenset(funsor_node['funsor']['log_prob'].inputs)
+                try:
+                    assert pyro_names == funsor_names
+                except AssertionError:
+                    assert pyro_names == frozenset(name.replace('__PARTICLES', '') for name in funsor_names)
+        except AssertionError:
+            for name, pyro_node in tr_pyro.nodes.items():
+                if pyro_node['type'] != 'sample':
+                    continue
+                funsor_node = tr_funsor.nodes[name]
+                pyro_names = frozenset(symbol_to_name[d] for d in pyro_node['packed']['log_prob']._pyro_dims)
+                funsor_names = frozenset(funsor_node['funsor']['log_prob'].inputs)
+                if pyro_names != funsor_names:
+                    err_str = "==> (packed mismatch) {}".format(name)
+                else:
+                    err_str = name
+                print(err_str, "Pyro: {} vs Funsor: {}".format(sorted(tuple(pyro_names)), sorted(tuple(funsor_names))))
+            raise
+
+    if _NAMED_TEST_STRENGTH >= 3:
+        try:
+            # finer check: exact match with unpacked Pyro shapes
```

Comment:
This stricter testing logic asserts that the unpacked shapes generated by `pyro.contrib.funsor` during enumeration for all value and log-probability tensors match those generated by Pyro exactly.  This was very useful early on in development, but since the implementation details of dimension allocation have changed somewhat it's no longer expected to hold for every test here and is rarely exercised.